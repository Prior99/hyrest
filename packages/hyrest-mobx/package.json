{
  "name": "hyrest-mobx",
  "version": "0.0.0",
  "description": "A hybrid REST library for the frontend and backend. Mobx connector package.",
  "keywords": [
    "hyrest",
    "hybrid",
    "http",
    "rest",
    "api",
    "promise",
    "form",
    "field",
    "mobx"
  ],
  "main": "dist/index.js",
  "author": "Frederick Gnodtke",
  "license": "MIT",
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf dist",
    "docs": "typedoc --mode file --out docs/ src/"
  },
  "typings": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/Prior99/hyrest"
  },
  "files": [
    "dist",
    "src",
    "LICENSE",
    "README.md"
  ],
  "devDependencies": {
    "typedoc": "^0.11.1",
    "typescript": "^3.0.1"
  },
  "dependencies": {
    "bind-decorator": "^1.0.11",
    "hyrest": "^0.8.0",
    "mobx": "^5.0.3",
    "reflect-metadata": "^0.1.12"
  }
}
