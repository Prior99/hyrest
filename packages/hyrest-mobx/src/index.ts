export * from "./field";
export * from "./decorators";
export * from "./types";
export * from "./validation-status";
