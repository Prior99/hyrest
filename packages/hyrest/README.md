# Hyrest

[![npm](https://img.shields.io/npm/v/hyrest.svg)](https://www.npmjs.com/package/hyrest)
[![Build Status](https://travis-ci.org/Prior99/hyrest.svg?branch=master)](https://travis-ci.org/Prior99/hyrest)
[![Coverage Status](https://coveralls.io/repos/github/Prior99/hyrest/badge.svg?branch=master)](https://coveralls.io/github/Prior99/hyrest?branch=master)

Hyrest is a hybrid REST framework for both the client and the server.

The idea is to define routes using decorators and use them to both serve the REST endpoint
and call them from the frontend. When developing both server and client in the same repository
or sharing a common library with all endpoints between the both, a call to a REST endoint
is transparent, type-safe and as easy calling a method.

## Usage

Please refer to the [resources](#resources) for getting you started.

## Resources

- [Tutorial](https://prior99.gitlab.io/hyrest/docs/tutorial-about)
- [Minimal example project](https://github.com/Prior99/hyrest-todo-example)
- [Documentation](https://prior99.gitlab.io/hyrest/)
- [Guide](https://prior99.gitlab.io/hyrest/docs/preamble-about/)
- [API Reference](https://prior99.gitlab.io/hyrest/api/hyrest/)
