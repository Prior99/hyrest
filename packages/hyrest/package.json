{
  "name": "hyrest",
  "version": "0.10.0",
  "description": "A hybrid REST library for the frontend and backend.",
  "keywords": [
    "hyrest",
    "hybrid",
    "http",
    "rest",
    "api",
    "promise"
  ],
  "main": "dist/index.js",
  "author": "Frederick Gnodtke",
  "license": "MIT",
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf dist",
    "docs": "typedoc --mode file --out docs/ src/"
  },
  "typings": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/Prior99/hyrest"
  },
  "files": [
    "dist",
    "src",
    "LICENSE",
    "README.md"
  ],
  "devDependencies": {
    "@types/invariant": "^2.2.29",
    "typedoc": "^0.12.0",
    "typescript": "^3.0.3"
  },
  "dependencies": {
    "http-status-codes": "^1.3.0",
    "invariant": "^2.2.2",
    "isomorphic-fetch": "^2.2.1",
    "path-to-regexp": "^2.4.0",
    "reflect-metadata": "^0.1.10"
  }
}
