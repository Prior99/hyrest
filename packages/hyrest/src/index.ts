import "path-to-regexp";

export * from "./all-keys";
export * from "./answers";
export * from "./authorization";
export * from "./configure";
export * from "./controller";
export * from "./converters";
export * from "./is-browser";
export * from "./last-call";
export * from "./parameters";
export * from "./processed";
export * from "./route";
export * from "./schema-generator";
export * from "./scope";
export * from "./transform";
export * from "./types";
export * from "./validation";
export * from "./validators";

import * as dataType from "./converters";
export const DataType = dataType;

import * as http from "./answers";
export const Http = http;
