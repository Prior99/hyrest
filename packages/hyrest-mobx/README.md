# Hyrest MobX

<img align="right" width="200" height="200" src="https://github.com/Prior99/hyrest/raw/master/logo/hyrest-logo-400px.png">

[![npm](https://img.shields.io/npm/v/hyrest-mobx.svg)](https://www.npmjs.com/package/hyrest-mobx)
[![pipeline status](https://gitlab.com/prior99/hyrest/badges/master/pipeline.svg)](https://github.com/Prior99/hyrest)
[![coverage report](https://gitlab.com/prior99/hyrest/badges/master/coverage.svg)](https://github.com/Prior99/hyrest)

Hyrest is a hybrid REST framework for both the client and the server.

**This is the MobX frontend utility package.**

After creating and exposing a REST Api using [hyrest](../hyrest) and [hyrest-express](../hyrest-express) it might be useful to re-use the existing models and validation from the server side in the client to store data next to forms and perform validation on it.

## Usage

Let's say a model has been defined which is already used by a controller:

```typescript
import { is, email, length } from "hyrest";
class User {
    @is().validate(email) public email: string;
    @is().validate(length(5, 20)) public name: string;
    @is().validate(length({ min: 10 })) public password: string;
}
```

In order to create a new user, one might want to implement a sign up form. This is where this package can come in handy.

Instead of creating a `@observable public email: string` and so on for each field on the `User` model, simple utilize hyrest-mobx:

```typescript
import * as React from "react";
import { observer } from "mobx-react";
import { field, hasFields, Field } from "hyrest-mobx";

@observer @hasFields()
class Signup extends React.Component {
    @field(User) private user: Field<User>;

    @action.bound private handleSubmit() {
        console.log(user.value); // Will log the unwrapped user.
    }

    public render() {
        return (
            <form onSubmit={this.handleSubmit}>
                <input type="text" {...this.user.nested.email.reactInput} />
                <input type="text" {...this.user.nested.name.reactInput} />
                <input type="password" {...this.user.nested.password.reactInput} />
                {
                    this.users.errors.map((err, index) => <p key={index}>{err}</p>)
                }
                <button disabled={!this.user.valid}>Signup</button>
            </form>
        );
    }
}
```

The decorator `@field` will mark the property as a field with the type `User`.

After marking the class with `@hasFields`, new `Field`s will automatically be injected upon creation.

The injected values already utilize Mobx under the hood and hence no `@observable`s are no longer necessary.

## Resources

- [Tutorial](https://prior99.gitlab.io/hyrest/docs/tutorial-about)
- [Minimal example project](https://github.com/Prior99/hyrest-todo-example)
- [Documentation](https://prior99.gitlab.io/hyrest/)
- [Guide](https://prior99.gitlab.io/hyrest/docs/preamble-about/)
- [API Reference](https://prior99.gitlab.io/hyrest/api/hyrest-mobx/)
