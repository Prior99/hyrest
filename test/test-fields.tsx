import * as React from "react";
import { mount } from "enzyme";
import { observer } from "mobx-react";
import { bind } from "bind-decorator";
import { schemaFrom, is, length, oneOf, email, DataType, required, range } from "hyrest";
import { field, hasFields, Field } from "hyrest-mobx";

class Pet {
    @is().validate(length(5, 10), required)
    public name: string;

    @is(DataType.int).validate(required, range(0, 10))
    public age: number;

    @is().validate(oneOf("cat", "dog"), required)
    public type: "cat" | "dog";
}

class User {
    @is().validate(length(5, 10), required)
    public name: string;

    @is().validate(email, required)
    public email: string;

    @is()
    public pet: Pet;
}

const ctx = {
    ownUser: User,
};

@observer @hasFields(() => ctx)
class UserForm extends React.Component {
    @field(User) private user: Field<User>;

    @bind private handleChange(event: React.SyntheticEvent<HTMLInputElement>) {
        this.user.nested.email.update((event.target as HTMLInputElement).value);
    }

    public render() {
        return (
            <div>
                <input value={this.user.nested.email.value} onChange={this.handleChange} />
                { this.user.nested.email.valid && <p>Valid!</p> }
                { this.user.nested.email.invalid && <p>Invalid!</p> }
                { this.user.nested.email.untouched && <p>Please type something!</p> }
                { this.user.nested.email.inProgress && <p>Loading...</p> }
            </div>
        );
    }
}

test("works as expected throughout a journey", async () => {
    const form = mount(<UserForm />);
    expect(form).toMatchSnapshot();

    form.find("input").simulate("change", { target: { value: "someone" }});
    expect(form).toMatchSnapshot();
    await new Promise(resolve => setTimeout(resolve));
    form.update();
    expect(form).toMatchSnapshot();

    form.find("input").simulate("change", { target: { value: "someone@example.com" }});
    expect(form).toMatchSnapshot();
    await new Promise(resolve => setTimeout(resolve));
    form.update();
    expect(form).toMatchSnapshot();
});

test("retrieving the constructed model from a nested field structure", async () => {
    @hasFields(() => ctx)
    class A {
        @field(User) public user: Field<User>;
    }

    const a = new A();
    a.user.nested.pet.nested.age.update(7);
    a.user.nested.email.update("someone@example.com");

    expect(a.user.value).toMatchSnapshot();
});
