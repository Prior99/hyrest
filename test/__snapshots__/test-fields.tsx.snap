// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`retrieving the constructed model from a nested field structure 1`] = `
User {
  "email": "someone@example.com",
  "pet": Pet {
    "age": 7,
  },
}
`;

exports[`updating with a nested field structure 1`] = `
User {
  "email": "someone@example.com",
  "name": "someone",
  "pet": Pet {
    "age": 7,
    "name": "Bonkers",
    "type": "dog",
  },
}
`;

exports[`validating a nested field structure 1`] = `
UserList {
  "users": Array [
    User {
      "email": "someone@example.com",
      "name": "someone",
      "pet": Pet {
        "age": 7,
        "name": "Bonkers",
        "type": "dog",
      },
    },
    User {
      "email": "anotherone@example.com",
      "name": "anotherone",
      "pet": Pet {
        "age": 3,
        "name": "Kitty",
        "type": "cat",
      },
    },
    User {
      "email": "third@example.com",
      "name": "thirdy",
      "pet": Pet {
        "age": 1,
        "name": "Bello",
        "type": "cat",
      },
    },
  ],
}
`;

exports[`validating a nested field structure 2`] = `
UserList {
  "users": Array [
    User {
      "email": "someone@example.com",
      "name": "someone",
      "pet": Pet {
        "age": 7,
        "name": "Bonkers",
        "type": "dog",
      },
    },
    User {
      "email": "invalid-email",
      "name": "anotherone",
      "pet": Pet {
        "age": 3,
        "name": "Kitty",
        "type": "cat",
      },
    },
    User {
      "email": "third@example.com",
      "name": "thirdy",
      "pet": Pet {
        "age": 1,
        "name": "Bello",
        "type": "cat",
      },
    },
  ],
}
`;

exports[`with an array in the structure 1`] = `
<UserListForm>
  <div>
    <ul />
    <input
      id="email"
      onChange={[Function]}
      value=""
    />
    <input
      id="name"
      onChange={[Function]}
      value=""
    />
    <button
      onClick={[Function]}
    >
      Add user to list
    </button>
  </div>
</UserListForm>
`;

exports[`with an array in the structure 2`] = `
<UserListForm>
  <div>
    <ul>
      <li
        key="someone@example.com"
      >
        someone
         (
        someone@example.com
        )
      </li>
    </ul>
    <input
      id="email"
      onChange={[Function]}
      value="someone@example.com"
    />
    <input
      id="name"
      onChange={[Function]}
      value="someone"
    />
    <button
      onClick={[Function]}
    >
      Add user to list
    </button>
  </div>
</UserListForm>
`;

exports[`works as expected throughout a journey 1`] = `
<UserForm>
  <div>
    <input
      onChange={[Function]}
      value=""
    />
    <p>
      Please type something!
    </p>
  </div>
</UserForm>
`;

exports[`works as expected throughout a journey 2`] = `
<UserForm>
  <div>
    <input
      onChange={[Function]}
      value="someone"
    />
    <p>
      Loading...
    </p>
  </div>
</UserForm>
`;

exports[`works as expected throughout a journey 3`] = `
<UserForm>
  <div>
    <input
      onChange={[Function]}
      value="someone"
    />
    <p>
      Invalid!
    </p>
  </div>
</UserForm>
`;

exports[`works as expected throughout a journey 4`] = `
<UserForm>
  <div>
    <input
      onChange={[Function]}
      value="someone@example.com"
    />
    <p>
      Loading...
    </p>
  </div>
</UserForm>
`;

exports[`works as expected throughout a journey 5`] = `
<UserForm>
  <div>
    <input
      onChange={[Function]}
      value="someone@example.com"
    />
    <p>
      Valid!
    </p>
  </div>
</UserForm>
`;
